lib/puppet/type.rb

```diff
@@ -1316,6 +1318,9 @@ def pathbuilder
       and the second run will log the edit made by Puppet.)"
 
     validate do |list|
+      if Puppet.settings[:strict] != :off
+        Puppet.deprecation_warning(_("You are attempting to use the `audit` metaparameter in %{resource}. This parameter is deprecated and will be ignored in a future release.") % { resource: @resource.path })
```

Comment:
So there's actually a method in `Puppet::Util::Logging` called `puppet_deprecation_warning` ostensibly for this purpose (though I couldn't find it actually used anywhere): https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/logging.rb#L145-L153.

>   Logs a warning whose origin comes from Puppet source rather than somewhere internal within Puppet.  Otherwise the same as deprecation_warning()

This allows us to specify the origin file/line for the deprecation, rather than printing this file/line of lib/puppet/type.rb. Luckily, `Puppet::Type` includes `Puppet::Util::Logging` on line 82 so this method is available. Also, both `line` and `file` are available as methods on `self` here and I think they'll do what you want.

So this can be revised to something like:

```ruby
puppet_deprecation_warning("The audit metaparameter is deprecated and will be ignored in a future release", {:line => line, :file => file})
```
Worth a shot anyway?